crates/trie/parallel/src/proof_task.rs

```diff
@@ -0,0 +1,319 @@
+//! and is responsible for managing the fixed number of database transactions created at the start
+//! of the task.
+//!
+//! Individual [`ProofTaskTx`] instances manage a dedicated [`InMemoryTrieCursorFactory`] and
+//! [`HashedPostStateCursorFactory`], which are each backed by a database transaction.
+
+use crate::root::ParallelStateRootError;
+use alloy_primitives::{map::B256Set, B256};
+use reth_db_api::{transaction::DbTx, Database};
+use reth_provider::{
+    providers::ConsistentDbView, BlockReader, DBProvider, DatabaseProviderFactory, ProviderResult,
+    StateCommitmentProvider,
+};
+use reth_trie::{
+    hashed_cursor::HashedPostStateCursorFactory, proof::StorageProof,
+    trie_cursor::InMemoryTrieCursorFactory, updates::TrieUpdatesSorted, HashedPostStateSorted,
+    StorageMultiProof,
+};
+use reth_trie_common::prefix_set::{PrefixSet, PrefixSetMut};
+use reth_trie_db::{DatabaseHashedCursorFactory, DatabaseTrieCursorFactory};
+use std::{
+    collections::VecDeque,
+    sync::{
+        mpsc::{channel, Receiver, Sender, TryRecvError},
+        Arc,
+    },
+    time::Instant,
+};
+use tokio::runtime::Handle;
+use tracing::debug;
+
+type StorageProofResult = Result<StorageMultiProof, ParallelStateRootError>;
+
+/// A task that manages sending multiproof requests to a number of tasks that have longer-running
+/// database transactions
+#[derive(Debug)]
+pub struct ProofTaskManager<Tx> {
+    /// Storage proof input pending execution
+    pending_proofs: VecDeque<(StorageProofInput, Sender<StorageProofResult>)>,
+    /// The underlying handle from which to spawn proof tasks
+    executor: Handle,
+    /// The proof task transactions, containing owned cursor factories that are reused for proof
+    /// calculation.
+    proof_task_txs: Vec<ProofTaskTx<Tx>>,
+    /// The currently running proof tasks.
+    running_proof_tasks: Vec<RunningProofTask<Tx>>,
+    /// A receiver for new proof tasks.
+    proof_task_rx: Receiver<ProofTaskMessage>,
+}
+
+impl<Tx> ProofTaskManager<Tx> {
+    /// Creates a new [`ProofTaskManager`] with the given max concurrency, creating that number of
+    /// cursor factories.
+    ///
+    /// Returns an error if the consistent view provider fails to create a read-only transaction.
+    pub fn new<Factory>(
+        executor: Handle,
+        view: ConsistentDbView<Factory>,
+        task_ctx: ProofTaskCtx,
+        max_concurrency: usize,
+        proof_task_rx: Receiver<ProofTaskMessage>,
+    ) -> ProviderResult<Self>
+    where
+        Factory: DatabaseProviderFactory<Provider: BlockReader, DB: Database<TX = Tx>>
+            + StateCommitmentProvider
+            + 'static,
+    {
+        // initialize proof task txs
+        let mut proof_task_txs = Vec::with_capacity(max_concurrency);
+        for item in &mut proof_task_txs {
+            let provider_ro = view.provider_ro()?;
+            let tx = provider_ro.into_tx();
+            *item = ProofTaskTx::new(tx, task_ctx.clone());
```

Comment:
we're currently paying for this upfront before twe spawn this, right?

which doesn't seem reasonable especially because we spawn this and can just do all of this in the background or on demand.

Reply:
Yeah, we could do this on the spawned `run` function

Reply:
We now create transactions on-demand